protected/humhub/modules/user/controllers/RegistrationController.php

```diff
@@ -91,29 +91,42 @@ public function actionIndex()
         } elseif (Yii::$app->session->has('authClient')) {
             $authClient = Yii::$app->session->get('authClient');
             $this->handleAuthClientRegistration($authClient, $registration);
-        } else {
-            Yii::warning('Registration failed: No token (query) or authclient (session) found!', 'user');
-            Yii::$app->session->setFlash('error', 'Registration failed.');
-            return $this->redirect(['/user/auth/login']);
         }
 
-        if ($registration->submitted('save') && $registration->validate() && $registration->register($authClient)) {
-            Yii::$app->session->remove('authClient');
+        if ($registration->submitted('save') && $registration->validate()) {
+            $existingUser = User::findIdentity($registration->getUser()->email);
+
+            if ($existingUser) {
+                // Log in the existing user
+                Yii::$app->user->login($existingUser);
```

Comment:
It seems before run the "login" action we should check for `User::STATUS_ENABLED` as it was implemented here and as it is implemented on the auth controller, so user sees a message like "Your account is not approved yet!".

Reply:
> I mean what if a user is already registered so `$existingUser !== null` but the user is not enabled (disabled or not approved). Will be the user logged in automatically without error message like "Your account is disabled!"?

I've tested with your concerns in mind, and can say that all this does is bypass the email steps and redirects the user to the account creation steps which if user approval is enabled then the user must wait till their account is approved, if the user email already exists they're able to connect their OAuth account and user account, the only issue that I'm seeing is that the even though the email is checked correctly the user is still redirected to the account creation step, so that is something to look into. 🤔

Reply:
I'd like to clarify the following;

- `$existingUser` redirects to > account creation > second click of oauth client button > logs in the user
- `$existingUser !== null` redirects to > account creation > normal registration process

Reply:
> * `$existingUser` redirects to > account creation > second click of oauth client button > logs in the user

Do you mean it works as you described even it the user is not active, i.e. when `$existingUser->status !== User::STATUS_ENABLED`?
Or such disabled user cannot be logged in?

Reply:
> > * `$existingUser` redirects to > account creation > second click of oauth client button > logs in the user
> 
> Do you mean it works as you described even it the user is not active, i.e. when `$existingUser->status !== User::STATUS_ENABLED`? Or such disabled user cannot be logged in?

From my tests if the registered user exists then they're logged in as normal but as stated on line 114 it makes sure that the status is `STATUS_ENABLED` for autologin, if the user is a newly registered user then from my tests the account isn't created but instead takes the user to the account creation step where then if admin approval is enabled the status is not used and the message for the user to wait for approval of the account.

Reply:
> From my tests if the registered user exists then they're logged in as normal

Did you test if the registered user has a "Disabled" or "Unapproved" status?
I think user should be restricted for log in action with such statuses, isn't?

Reply:
> > From my tests if the registered user exists then they're logged in as normal
> 
> Did you test if the registered user has a "Disabled" or "Unapproved" status? I think user should be restricted for log in action with such statuses, isn't?

From my testing the user isn't created through this method it bypasses the process of requiring the go through the email token to get to the account creation section as stated, once the account creation form is filled out you're given the following screen;

![Screenshot_1](https://github.com/humhub/humhub/assets/35392110/4405915d-9efc-4615-bcb8-87b7917c862d)

Till the account is approved the user can't login if they're a new user, if the auth client doesn't properly handle `defaultNormalizeUserAttributeMap()` then no email will be displayed when the user registers their account, here's an example of what happens with approval needed by an admin;

![Screenshot_2](https://github.com/humhub/humhub/assets/35392110/5dda6960-9e07-46f9-b7df-989788767612)

The only issue is, if the registering user tries logging in they're redirected back to the account creation page, which I've not found a way of fixing yet... 🤔

Reply:
> The only issue is, if the registering user tries logging in they're redirected back to the account creation page, which I've not found a way of fixing yet... 🤔

I think if user is registered but it cannot be logged in by some reason (maybe it is disabled or not approved) then user should see a message about the reason as it is displayed when user is logged in by login/pass - https://github.com/humhub/humhub/blob/master/protected/humhub/modules/user/controllers/AuthController.php#L287-L292

Reply:
> > The only issue is, if the registering user tries logging in they're redirected back to the account creation page, which I've not found a way of fixing yet... 🤔
> 
> I think if user is registered but it cannot be logged in by some reason (maybe it is disabled or not approved) then user should see a message about the reason as it is displayed when user is logged in by login/pass - https://github.com/humhub/humhub/blob/master/protected/humhub/modules/user/controllers/AuthController.php#L287-L292



> > The only issue is, if the registering user tries logging in they're redirected back to the account creation page, which I've not found a way of fixing yet... 🤔
> 
> I think if user is registered but it cannot be logged in by some reason (maybe it is disabled or not approved) then user should see a message about the reason as it is displayed when user is logged in by login/pass - https://github.com/humhub/humhub/blob/master/protected/humhub/modules/user/controllers/AuthController.php#L287-L292



> > The only issue is, if the registering user tries logging in they're redirected back to the account creation page, which I've not found a way of fixing yet... 🤔
> 
> I think if user is registered but it cannot be logged in by some reason (maybe it is disabled or not approved) then user should see a message about the reason as it is displayed when user is logged in by login/pass - https://github.com/humhub/humhub/blob/master/protected/humhub/modules/user/controllers/AuthController.php#L287-L292

That is where the original issue came into play, too many else statements ended up causing the following to always return as `true`

```php
else {	
            Yii::warning('Registration failed: No token (query) or authclient (session) found!', 'user');	
            Yii::$app->session->setFlash('error', 'Registration failed.');	
            return $this->redirect(['/user/auth/login']);
}
```

Reply:
ok, if you think it works as expected now then I could approve it, because I cannot test it on my local server now.
I will test this later on community server because I have similar issue with my account here https://github.com/humhub/humhub-internal/issues/258.

Reply:
> ok, if you think it works as expected now then I could approve it, because I cannot test it on my local server now. I will test this later on community server because I have similar issue with my account here https://github.com/humhub/humhub-internal/issues/258.

Maybe the issue could be more simpler than we think? Looking at how we're getting the token and how it's validated could also be the reason why the error is happening? There are so many other possibilities that could be the root cause of it;
https://github.com/humhub/humhub/blob/ef41e8ab37b5a810e96815ac590ab41a2340a840/protected/humhub/modules/user/services/InviteRegistrationService.php#L41-L44

I've tested to see if `$token` is correct with the following and it is correct but I'm not sure if the validation is correct in the if statement `$inviteRegistrationService->isValid()`;
```php
// Log the token value
$token = Yii::$app->request->get('token');
Yii::info('Token value: ' . var_export($token, true), 'debug');

if ($token) {
    $inviteRegistrationService = new InviteRegistrationService($token);
    Yii::info('Token is present, checking validity.', 'debug');
    if (!$inviteRegistrationService->isValid()) {
        Yii::error('Invalid registration token!', 'user');
        throw new HttpException(404, 'Invalid registration token!');
    }
    $inviteRegistrationService->populateRegistration($registration);
} else {
    Yii::info('Token is not present.', 'debug');
}

if (Yii::$app->session->has('authClient')) {
    $authClient = Yii::$app->session->get('authClient');
    Yii::info('AuthClient found in session: ' . var_export($authClient, true), 'debug');
    $this->handleAuthClientRegistration($authClient, $registration);
} else {
    Yii::info('AuthClient not found in session.', 'debug');
}

if (!$token && !Yii::$app->session->has('authClient')) {
    Yii::warning('Registration failed: No token (query) or authClient (session) found!', 'user');
    Yii::$app->session->setFlash('error', 'Registration failed.');
    return $this->redirect(['/user/auth/login']);
}
```